### DISCUSSION TITLE
ESP32-WROOM-32E + PZEM004T

### DISCUSSION THREAD

**Discussion by dexterbot80** [arendst/Tasmota#18296] [Q&A]
Hello,
I want to connect [6x PZEM004T](https://www.aliexpress.com/i/32707315880.html) energy sensors to [ESP32-WROOM-32E](https://www.olimex.com/Products/IoT/ESP32/ESP32-POE/open-source-hardware) to monitor two different three-phase energy sources.
I use the [ESP32-POE board from OLIMEX](https://www.olimex.com/Products/IoT/ESP32/ESP32-POE/open-source-hardware) and the [Peacefair-PZEM-004T sensors](https://www.aliexpress.com/i/32707315880.html).
I can't add each sensor with a different address like on the ESP8266 - I get the command unknown error.
`21:28:40.084 CMD: ModuleAddress 1
21:28:40.088 RSL: RESULT = {"Command":"Error"}`
Likewise on the webinterface, the read values keep disappearing and reappearing every few seconds...
Can someone help me configure the 6 sensors on the ESP32?

Thank you!

**Comment by barbudor** [comment]
Generaly getting "Command: Error" means that something is wrongly configured
I suggest you read this How-To https://github.com/arendst/Tasmota/discussions/10567
Especially how the GPIO are configured for multiple module : PZEM016RX and not PZEM004RX

**Reply by dexterbot80** [reply]
> Generaly getting "Command: Error" means that something is wrongly configured I suggest you read this How-To #10567 Especially how the GPIO are configured for multiple module : PZEM016RX and not PZEM004RX

Hello and thank you for your help,
The PZEM004T circuit is [modified](https://raw.githubusercontent.com/TrDA-hab/Projects/master/PZEM%2BESP-01S/PZEM-mod2.jpg) according to the instructions in the link you sent me and now I installed the Tasmota32 LVGL using the web installer.

I'm testing for the first time with a single sensor because I'm encountering some problems.

I think I should solve these problems first and then add more sensors.
Please see the attached video and tell me what mistakes I made here and please also look carefully at the total energy consumed after I reconnect the power supply of the PZEM004T sensor.
[https://drive.google.com/file/d/1FxKaXipildJ_436-Guk3xh-YyGTjVMJV/view?usp=share_link](https://drive.google.com/file/d/1FxKaXipildJ_436-Guk3xh-YyGTjVMJV/view?usp=share_link)
Thank you

**Comment by dexterbot80** [comment]
Hello,
In the end I managed to make it work but only on the esp8266.  [https://drive.google.com/file/d/1wM9JWsqHgqFtKC5wL_SqDM0wQ79D6nqP/view?usp=share_link](https://drive.google.com/file/d/1wM9JWsqHgqFtKC5wL_SqDM0wQ79D6nqP/view?usp=share_link)
But I don't understand what I have to do to make it work on the esp32.
[https://drive.google.com/file/d/1hRJ6JfdPF_Meq7jgUwTrHNGQV_iFFs0r/view?usp=share_link](https://drive.google.com/file/d/1hRJ6JfdPF_Meq7jgUwTrHNGQV_iFFs0r/view?usp=share_link)
I compile simultaneously for both esp32 and esp8266 and if I install the tasmota32.bin file on the esp32 it does not work correctly.
The measured values are displayed intermittently, eg: Voltage 230V for 2 sec Voltage 0V for 10 sec Voltage 230V for 2 sec ..............and so on.
Do I have to modify something to make it work the same as on the esp8266?
Thank you

**Reply by arendst** [reply]
Did you use latest dev branch for the ESP32? If not, do so as it contains fixes in that area.

**Reply by dexterbot80** [reply]
> Did you use latest dev branch for the ESP32? If not, do so as it contains fixes in that area.

Hello Theo,
Thank you for the suggestion. Do you mean the version you are working on now?
This one? [https://github.com/arendst/Tasmota](https://github.com/arendst/Tasmota)
I used the source code version 12.4.0 Peter.
[https://github.com/arendst/Tasmota/releases/tag/v12.4.0](https://github.com/arendst/Tasmota/releases/tag/v12.4.0)
I want to use esp32 because I found some boards that are equipped with ethernet and I would like to connect 6 x PZEM004T to monitor 2 three-phase power sources simultaneously.
I'm going to install the device in a grounded metal electric panel and I don't think it would work WIFI because of the Faraday shield :)

Thanks again

**Reply by Amruth-Smaragdine** [reply]
Please explain the steps you followed to make it work on ESP8266.

**Reply by barbudor** [reply]
Do you get that the title of this discussion is ESP32 + PZEM ?
If you have problems with ESP8266 don't hijack and create a new discussion starting to explain what you have done and where it start not working

**Comment by dexterbot80** [comment]
Here i tested using [Tasmota-master  branch](https://github.com/arendst/Tasmota/tree/master) and it seems to work roughly the same as version 12.4.0 Peter from this point of view
https://drive.google.com/file/d/1d7kGDJZ8QhtpIHKt4aEim25WJFkCP0Y2/view?usp=share_link

**Reply by barbudor** [reply]
Master = release

Use the development branch or a pre-built from http://ota.tasmota.com/tasmota32/

**Reply by dexterbot80** [reply]
> Master = release
> 
> Use the development branch or a pre-built from http://ota.tasmota.com/tasmota32/

Hello, First i install [tasmota32-LVGL.bin](http://ota.tasmota.com/tasmota32/tasmota32-lvgl.bin) and [tasmota32.bin](http://ota.tasmota.com/tasmota32/tasmota32.bin) and it does not work correctly with pzem004T
Now we have downloaded the version in development, let's see what happens.

**Reply by dexterbot80** [reply]
> Master = release
> 
> Use the development branch or a pre-built from http://ota.tasmota.com/tasmota32/
I hardly found the cherry on the cake :)
**I confirm that the error found earlier is resolved in the in development version** [12.4.0.4](https://github.com/arendst/Tasmota/tree/development)
Here is the test video 
https://drive.google.com/file/d/1g0ALW48tXwqKKJaoBAYIhcUvTKhBM5Zx/view?usp=share_link

I also wrote a rule for an action in case of minimum or maximum voltage.
Can someone confirm if I wrote the rule correctly? 

First time we defined the minimum and maximum voltage using the expressions:
```
mem1 195; 
mem2 245;
```
And I wrote the rule like this:

```
Rule1
ON Tele-ENERGY#Voltage<=%mem1% DO Power1 1 BREAK
ON Tele-ENERGY#Voltage>%mem1% DO Power1 0 ENDON
ON Tele-ENERGY#Voltage<=%mem2% DO Power1 0 BREAK
ON Tele-ENERGY#Voltage>%mem2% DO Power1 1 ENDON
```

Thank you very much guys, you are doing a great job here

**Reply by barbudor** [reply]
While syntaxically correct, I don't think the rule will do what you want
I would suggest to try 
```
Rule1
   ON Tele-ENERGY#Voltage<=%mem1% DO Power1 1 BREAK
   ON Tele-ENERGY#Voltage>%mem2% DO Power1 1 BREAK
   ON Tele-Energy#Voltage DO Power1 0 ENDON
```

**Reply by dexterbot80** [reply]
Thank you #barbudor
My rule does exactly the same thing, but the rule you wrote is much simpler. 
I didn't think it could be that simple here:
`ON Tele-Energy#Voltage DO Power1 0 ENDON`
I used an autotransformer to simulate the electrical network and can set any alternative voltage between 0 and 260V.

**Comment by smcgann99** [comment]
I also had some issues with 2 pzems connected to esp32. Despite connecting 1 at a time and using moduleaddress 1 and 2. When both were connected they both answer to the same address, as both TX flash. I downgraded from 12.4.0.5 to 11.1.0.

I was able to address them and get them working then. After upgrading back to 12.4.0.5 they continued to work, so I think the issue is with the moduleaddress command not working correctly.

I got the idea to downgrade after seeing some else had issues on discord -

**_alex77
 — 
Yesterday at 15:57
Hello! is it possibile that Tasmota 12.4 does not support multiple (2) PZEM-004 v1? Downgraded to v 11.1 on same hw config and eveithing worked fine._**

**Reply by arendst** [reply]
I love questions where users quote the answer....

**Reply by smcgann99** [reply]
> I love questions where users quote the answer....

But what is the answer ? I couldn't see any open issues or pull request about this, so I was reporting my findings with latest dev version, also adding that it "seems" like the problem is with the moduleaddress command, since 12.4.05 on ESP32 does work, but only if I set the moduleaddress using an older version.

**Comment by Amruth-Smaragdine** [comment]
After I referred to many articles, videos, and blogs, I was able to make it work. I have captured the complete information to a few INSTRUCTABLES 
https://www.instructables.com/member/AmruthKumarSala/instructables/
I also have a few YouTube Videos in this playlist: https://www.youtube.com/watch?v=3L-hY_tTx24&list=PL0eiemFx0ClTO9qRJ8L540a3W-12NzTdY
Comments & constructive criticism to improve the content are appreciated.
